### PR TITLE
S2bot

### PR DESCRIPTION
**Pull request by achipa**
S2Bot-based input tracker for embedded sensor boards (ie BBC micro:bit)

### CODE REVIEW

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+				qDebug() << "Request submitted OK";
+			}
+			else {
+				qWarning() << "Request bounced:" << reply->attribute(QNetworkRequest::HttpStatusCodeAttribute) << reply->errorString();
+				return;
+			}
+			QByteArray ba = reply->readAll();
+			QStringList slist = QString(ba).split(QRegExp("[\r\n]"), QString::SkipEmptyParts);
+			int order[] = {
+				bound<int>(s.idx_x, 0, 3),
```
Comment (sthalik) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: That's an off-by-one -- you want "2" as the maximum index.

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        flag_Raw = 1 << 0,
+        flag_Orient = 1 << 1,
+        Mask = flag_Raw | flag_Orient
+    };
+
+	if (s.freq == 0) s.freq = 10;
+	timer.setInterval(1000.0/s.freq);
+	timer.setSingleShot(false);	
+	connect(&timer, &QTimer::timeout, [this]() {
+		auto reply = m_nam->get(QNetworkRequest(QUrl("http://localhost:17317/poll")));
```
Comment (sthalik) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: A http request adds significant overhead. Please note that camera tracking does 187 Hz with no worry, and IMU tracker does 250 Hz, same as all Oculus Rift versions. Have you tried it in some flight/other simulation software? How bad is it?

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		connect(reply, &QNetworkReply::finished, [this, reply]() {
```
Comment (sthalik) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: You should specify `Qt::DirectConnection` just to be safe. Potentially dispatching timers via the event queue is expensive. Your timer function is thread-safe already.

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+			if (reply->error() == QNetworkReply::NoError) {
+				qDebug() << "Request submitted OK";
+			}
+			else {
+				qWarning() << "Request bounced:" << reply->attribute(QNetworkRequest::HttpStatusCodeAttribute) << reply->errorString();
+				return;
+			}
+			QByteArray ba = reply->readAll();
+			QStringList slist = QString(ba).split(QRegExp("[\r\n]"), QString::SkipEmptyParts);
```
Comment (sthalik) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: Try using `[\r\n]+` but I nitpick here.

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+{
+    if (datum < least)
+        return least;
+    if (datum > max)
+        return max;
+    return datum;
+}
+
+void tracker_s2bot::run() {
+#pragma pack(push, 1)
```
Comment (sthalik) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: This is dead code

File: tracker-s2bot/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# S2Bot plugin for opentrack
+This is a tracker providing 3D head tracking using the S2Bot helper application. 
+
+S2Bot is originally a plugin that connect embedded devices with the [Scratch programming environment](https://scratch.mit.edu/). This plugin emulates the Scratch interface to get access to the sensor data from the various boards. It was developed and tested using the [micro:bit](https://www.microbit.co.uk) board, but should work with any of the S2Bot supported sensor boards that have accelerometers and/or magnetometers. S2Bot is available for Windows, Mac and Linux.
```
Comment (sthalik) [inline tracker-s2bot/README.md]: Smalltalk :+1: :cat2:

File: tracker-s2bot/ftnoir_tracker_s2bot.cpp (hunk 2 above)
Comment (achipa) [inline tracker-s2bot/ftnoir_tracker_s2bot.cpp]: Http is unfortunately a given as that's what S2Bot exposes.

I use it with Elite:Dangerous (a space sim), and while there is some lag and it's far from Oculus level performance, I find it usable (especially considering the Micro:Bit is a 10$ board that kids get for free). 

The call frequency can be of course raised (it's actually one of the configurable parameters), but I did not experience significant difference over 30Hz (note that S2Bot or the sensor boards themselves might be the bottleneck there).

**Comment by sthalik** [comment]
Thanks!

**Comment by sthalik** [comment]
@achipa I'd like you to put your own copyright statement into the file. I haven't written any substantial parts of the module so it doesn't belong to me in any case.

I'd rather you convert the code from tabs to 4 spaces. It shows the authorship on `git blame`. Changing the copyright is just enough though.

**Pull request merged**